File: java/e2e/src/test/java/org/apache/rocketmq/broker/server/abnormal/PushConsumerRetryTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.broker.server.abnormal;

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Vector;
import java.util.concurrent.Callable;
import java.util.concurrent.ConcurrentHashMap;
import org.apache.rocketmq.client.apis.ClientException;
import org.apache.rocketmq.client.apis.consumer.ConsumeResult;
import org.apache.rocketmq.client.apis.consumer.FilterExpression;
import org.apache.rocketmq.client.apis.consumer.FilterExpressionType;
import org.apache.rocketmq.client.apis.consumer.PushConsumer;
import org.apache.rocketmq.client.apis.consumer.SimpleConsumer;
import org.apache.rocketmq.client.apis.message.Message;
import org.apache.rocketmq.client.apis.message.MessageView;
import org.apache.rocketmq.client.rmq.RMQNormalProducer;
import org.apache.rocketmq.common.attribute.TopicMessageType;
import org.apache.rocketmq.enums.TESTSET;
import org.apache.rocketmq.factory.ClientConfigurationFactory;
import org.apache.rocketmq.factory.ConsumerFactory;
import org.apache.rocketmq.factory.MessageFactory;
import org.apache.rocketmq.factory.ProducerFactory;
import org.apache.rocketmq.frame.BaseOperate;
import org.apache.rocketmq.util.NameUtils;
import org.apache.rocketmq.util.RandomUtils;
import org.apache.rocketmq.util.VerifyUtils;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@Tag(TESTSET.DELAY)
public class PushConsumerRetryTest extends BaseOperate {
    private final Logger log = LoggerFactory.getLogger(PushConsumerRetryTest.class);
    private String tag;
    private final static int SEND_NUM = 10;

    @BeforeEach
    public void setUp() {
        tag = NameUtils.getRandomTagName();
    }

    @Test
    @DisplayName("The normal message is sent, and after the PushConsumer partial retry, the retry message is expected to be consumed")
    public void testNormalTopicPushConsumerRetry() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String topic = getTopic(TopicMessageType.NORMAL.getValue(), methodName);
        String groupId = getGroupId(methodName);
        RMQNormalProducer producer = ProducerFactory.getRMQProducer(account, topic);
        Map<String, MessageView> noRetryMsgs = new ConcurrentHashMap<>();
        Map<String, MessageView> retryMsgs = new ConcurrentHashMap<>();

        SimpleConsumer simpleConsumer = ConsumerFactory.getSimpleConsumer(account, topic, groupId, new FilterExpression(tag), Duration.ofSeconds(10));
        VerifyUtils.tryReceiveOnce(simpleConsumer);

        try {
            provider.newPushConsumerBuilder().setClientConfiguration(ClientConfigurationFactory.build(account)).setConsumerGroup(groupId).setSubscriptionExpressions(Collections.singletonMap(topic, new FilterExpression(tag, FilterExpressionType.TAG))).setConsumptionThreadCount(20).setMessageListener(messageView -> {
                if (Integer.parseInt(StandardCharsets.UTF_8.decode(messageView.getBody()).toString()) % 2 == 0) {
                    log.info(String.format("commit normal msg %s ", messageView));
                    noRetryMsgs.putIfAbsent(messageView.getMessageId().toString(), messageView);
                    return ConsumeResult.SUCCESS;
                } else if (messageView.getDeliveryAttempt() == 2) {
                    log.info(String.format("retry  normal msg %s ", messageView));
                    retryMsgs.putIfAbsent(messageView.getMessageId().toString(), messageView);
                    return ConsumeResult.SUCCESS;
                }
                return ConsumeResult.FAILURE;
            }).build();
        } catch (ClientException e) {
            throw new RuntimeException(e);
        }
        Assertions.assertNotNull(producer, "Get Producer Failed");
        for (int i = 0; i < SEND_NUM; i++) {
            Message message = MessageFactory.buildMessage(topic, tag, String.valueOf(i));
            producer.send(message);
        }
        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");
        //All messages are consumed.
        await().atMost(120, SECONDS).until(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return noRetryMsgs.size() == retryMsgs.size() && noRetryMsgs.size() == SEND_NUM / 2;
            }
        });
        for (Object msgId : producer.getEnqueueMessages().getAllData()) {
            Assertions.assertTrue(noRetryMsgs.containsKey(msgId) || retryMsgs.containsKey(msgId));
        }
    }

    @Test
    @DisplayName("The send order message, after the PushConsumer partial retry, is expected to consume the retry message, and the message consumption order and send order")
    public void testFiFoTopicPushConsumerRetry() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String topic = getTopic(TopicMessageType.FIFO.getValue(), methodName);
        String groupId = getOrderlyGroupId(methodName);

        SimpleConsumer simpleConsumer = ConsumerFactory.getSimpleConsumer(account, topic, groupId, new FilterExpression(tag), Duration.ofSeconds(10));
        VerifyUtils.tryReceiveOnce(simpleConsumer);

        PushConsumer pushConsumer = null;
        RMQNormalProducer producer = ProducerFactory.getRMQProducer(account, topic);
        Assertions.assertNotNull(producer);
        Vector<MessageView> recvMessages = new Vector<>();
        try {
            pushConsumer = provider.newPushConsumerBuilder()
                .setClientConfiguration(ClientConfigurationFactory.build(account))
                .setConsumerGroup(groupId)
                .setSubscriptionExpressions(Collections.singletonMap(topic, new FilterExpression(tag)))
                .setConsumptionThreadCount(20)
                .setMessageListener(messageView -> {
                    if (Integer.parseInt(StandardCharsets.UTF_8.decode(messageView.getBody()).toString()) % 2 == 0) {
                        log.info(String.format("commit fifo msg %s ", messageView));
                        recvMessages.add(messageView);
                        return ConsumeResult.SUCCESS;
                    } else if (messageView.getDeliveryAttempt() == 2) {
                        log.info(String.format("retry fifo msg %s ", messageView));
                        recvMessages.add(messageView);
                        return ConsumeResult.SUCCESS;
                    }
                    return ConsumeResult.FAILURE;
                }).build();
        } catch (ClientException e) {
            throw new RuntimeException(e);
        }
        Assertions.assertNotNull(producer, "Get Producer Failed");
        String messageGroup = RandomUtils.getStringByUUID();
        for (int i = 0; i < SEND_NUM; i++) {
            Message message = MessageFactory.buildOrderMessage(topic, tag, String.valueOf(i), messageGroup);
            producer.send(message);
        }
        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");
        //All messages are consumed.
        await().atMost(120, SECONDS).until(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return recvMessages.size() == SEND_NUM;
            }
        });
        for (int i = 0; i < SEND_NUM; i++) {
            Assertions.assertEquals(i, Integer.parseInt(StandardCharsets.UTF_8.decode(recvMessages.get(i).getBody()).toString()), "recv message failed");
        }
    }

    @Test
    @DisplayName("Send sequential messages, using three Shardingkeys, after partial retries, expect to consume retry messages, and the order of message consumption and the order of message delivery")
    public void testFiFoTopicWith3ShardingkeyPushConsumerRetry() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String topic = getTopic(TopicMessageType.FIFO.getValue(), methodName);
        String groupId = getOrderlyGroupId(methodName);

        SimpleConsumer simpleConsumer = ConsumerFactory.getSimpleConsumer(account, topic, groupId, new FilterExpression(tag), Duration.ofSeconds(10));
        VerifyUtils.tryReceiveOnce(simpleConsumer);

        PushConsumer pushConsumer = null;
        RMQNormalProducer producer = ProducerFactory.getRMQProducer(account, topic);
        Assertions.assertNotNull(producer);
        Vector<MessageView> recvMessages = new Vector<>();
        try {
            pushConsumer = provider.newPushConsumerBuilder()
                .setClientConfiguration(ClientConfigurationFactory.build(account))
                .setConsumerGroup(groupId)
                .setSubscriptionExpressions(Collections.singletonMap(topic, new FilterExpression(tag)))
                .setConsumptionThreadCount(20)
                .setMessageListener(messageView -> {
                    if (Integer.parseInt(StandardCharsets.UTF_8.decode(messageView.getBody()).toString()) % 2 == 0) {
                        log.info(String.format("commit fifo msg %s ", messageView));
                        recvMessages.add(messageView);
                        return ConsumeResult.SUCCESS;
                    } else if (messageView.getDeliveryAttempt() == 2) {
                        log.info(String.format("retry fifo msg %s ", messageView));
                        recvMessages.add(messageView);
                        return ConsumeResult.SUCCESS;
                    }
                    return ConsumeResult.FAILURE;
                }).build();
        } catch (ClientException e) {
            throw new RuntimeException(e);
        }
        Assertions.assertNotNull(producer, "Get Producer Failed");
        for (int i = 0; i < SEND_NUM; i++) {
            Message message = MessageFactory.buildOrderMessage(topic, tag, String.valueOf(i), String.valueOf(i % 3));
            producer.send(message);
        }
        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");
        //All messages are consumed.
        await().atMost(180, SECONDS).until(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                return recvMessages.size() == SEND_NUM;
            }
        });

        ArrayList<MessageView> shardingKeyA = new ArrayList<MessageView>();
        ArrayList<MessageView> shardingKeyB = new ArrayList<MessageView>();
        ArrayList<MessageView> shardingKeyC = new ArrayList<MessageView>();
        for (int i = 0; i < SEND_NUM; i++) {
            MessageView messageView = recvMessages.get(i);
            switch (messageView.getMessageGroup().toString()) {
                case "Optional[a]":
                    shardingKeyA.add(messageView);
                    break;
                case "Optional[b]":
                    shardingKeyB.add(messageView);
                    break;
                case "Optional[c]":
                    shardingKeyC.add(messageView);
                    break;
                default:
                    log.info(String.format("recv msgs with wrong shardingkey: %s, %s",
                        messageView.getMessageGroup(), messageView));
            }
        }
        Assertions.assertTrue(isSorted(shardingKeyA));
        Assertions.assertTrue(isSorted(shardingKeyB));
        Assertions.assertTrue(isSorted(shardingKeyC));
    }

    public static boolean isSorted(List<MessageView> messageViewList) {
        for (int i = 0; i < messageViewList.size() - 1; i++) {
            if (Integer.parseInt(StandardCharsets.UTF_8.decode(messageViewList.get(i).getBody()).toString())
                > Integer.parseInt(StandardCharsets.UTF_8.decode(messageViewList.get(i + 1).getBody()).toString())) {
                return false;
            }
        }
        return true;
    }

}

File: cpp/rocketmq-client-cpp-tests/cpp4.x/test/server/abnormal/PushConsumerRetryTest.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <chrono>
#include <iostream>
#include <cassert>
#include <string>
#include "gtest/gtest.h"
#include "spdlog/logger.h"
#include "rocketmq/MQMessageExt.h"
#include "rocketmq/DefaultMQPushConsumer.h"
#include "rocketmq/MQMessageListener.h"
#include "enums/MessageType.h"
#include "frame/BaseOperate.h"
#include "resource/Resource.h"
#include "utils/NameUtils.h"
#include "utils/RandomUtils.h"
#include "utils/VerifyUtils.h"
#include "utils/SimpleConcurrentHashMapUtils.h"
#include "utils/SimpleConcurrentVectorUtils.h"
#include "factory/ConsumerFactory.h"
#include "factory/ProducerFactory.h"
#include "factory/MessageFactory.h"

extern std::shared_ptr<spdlog::logger> multi_logger;
extern std::shared_ptr<Resource> resource;

class ExceptionMsgListener : public rocketmq::MessageListenerConcurrently {
public:
    SimpleConcurrentHashMap<std::string, rocketmq::MQMessageExt> firstMsgs;
    SimpleConcurrentHashMap<std::string, rocketmq::MQMessageExt> retryMsgs;
    ExceptionMsgListener() {}
    virtual ~ExceptionMsgListener() {}

  virtual rocketmq::ConsumeStatus consumeMessage(const std::vector<rocketmq::MQMessageExt>& msgs) {
    for (size_t i = 0; i < msgs.size(); ++i) {
        auto msg = msgs[i];
        if(msg.getReconsumeTimes() == 1){
            if(!retryMsgs.contains(msg.getMsgId())){
                retryMsgs.insert(msg.getMsgId(), msg);
            }
            multi_logger->info("consume success: {}",msg.getMsgId());
        }else{
            // Simulate consuming operations
            multi_logger->info("{}","Simulate consuming operations fail");
            throw std::runtime_error("Simulate consuming operations fail");
            multi_logger->info("{}","Simulate consuming operations fail end");
            if(!firstMsgs.contains(msg.getMsgId())){
                firstMsgs.insert(msg.getMsgId(), msg);
            }
            multi_logger->info("recv msg(fail) {} ", msg.getMsgId());
        }
    }
    return rocketmq::CONSUME_SUCCESS;
  }
};

// class NullMsgListener : public rocketmq::MessageListenerConcurrently {
// public:
//     SimpleConcurrentHashMap<std::string, rocketmq::MQMessageExt> firstMsgs;
//     SimpleConcurrentHashMap<std::string, rocketmq::MQMessageExt> retryMsgs;
//     NullMsgListener() {}
//     virtual ~NullMsgListener() {}

//   virtual rocketmq::ConsumeStatus consumeMessage(const std::vector<rocketmq::MQMessageExt>& msgs) {
//     for (size_t i = 0; i < msgs.size(); ++i) {
//         auto msg = msgs[i];
//         if(msg.getReconsumeTimes() == 2){
//             if(!retryMsgs.contains(msg.getMsgId())){
//                 retryMsgs.insert(msg.getMsgId(), msg);
//             }
//             multi_logger->info("consume success: {}",msg.getMsgId());
//         }else{
//             // Simulate consuming operations
//             multi_logger->info("{}","Simulate consuming operations return null");
//             if(!firstMsgs.contains(msg.getMsgId())){
//                 firstMsgs.insert(msg.getMsgId(), msg);
//             }
//             multi_logger->info("recv msg(fail) {} ", msg.getMsgId());
//             return nullptr;
//         }
//     }
//     return rocketmq::CONSUME_SUCCESS;
//   }
// };

class NormalMsgListener : public rocketmq::MessageListenerConcurrently {
public:
    SimpleConcurrentHashMap<std::string, rocketmq::MQMessageExt> firstMsgs;
    SimpleConcurrentHashMap<std::string, rocketmq::MQMessageExt> retryMsgs;
    NormalMsgListener() {}
    virtual ~NormalMsgListener() {}

  virtual rocketmq::ConsumeStatus consumeMessage(const std::vector<rocketmq::MQMessageExt>& msgs) {
    for (size_t i = 0; i < msgs.size(); ++i) {
        auto msg = msgs[i];
        if(msg.getReconsumeTimes() > 0){
            if(!retryMsgs.contains(msg.getMsgId())){
                retryMsgs.insert(msg.getMsgId(), msg);
            }
            multi_logger->info("consume success: {}",msg.getMsgId());
            return rocketmq::CONSUME_SUCCESS;
        }else{
            if(!firstMsgs.contains(msg.getMsgId())){
                firstMsgs.insert(msg.getMsgId(), msg);
            }
            multi_logger->info("recv msg(fail) {} ", msg.getMsgId());
            return rocketmq::RECONSUME_LATER;
        }
    }
    return rocketmq::CONSUME_SUCCESS;
  }
};

class FIFOMsgListener : public rocketmq::MessageListenerOrderly {
public:
    SimpleConcurrentVector<rocketmq::MQMessageExt> recvMessages;
    FIFOMsgListener() {}
    virtual ~FIFOMsgListener() {}

  virtual rocketmq::ConsumeStatus consumeMessage(const std::vector<rocketmq::MQMessageExt>& msgs) {
    for (size_t i = 0; i < msgs.size(); ++i) {
        auto msg = msgs[i];
        if(msg.getReconsumeTimes() > 0){
            multi_logger->info("consume success: {}",msg.getMsgId());
            recvMessages.push_back(msg);
            return rocketmq::CONSUME_SUCCESS;
        }else{
            multi_logger->info("recv msg(fail) {} ", msg.getMsgId());
            return rocketmq::RECONSUME_LATER;
        }
    }
    return rocketmq::CONSUME_SUCCESS;
  }
};

//Simulate pushconsumer consumption fail, expect that the original message was not received, and capture all messages after message retry
TEST(PushConsumerRetryTest, testExceptionConsumption){
    int SEND_NUM = 5;
    std::string topic = getTopic(MessageType::NORMAL, "testExceptionConsumption", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("testExceptionConsumption");
    std::string tag = NameUtils::getRandomTagName();

    ExceptionMsgListener* listener = new ExceptionMsgListener();
    auto rmqPushConsumer = std::make_shared<rocketmq::DefaultMQPushConsumer>(group);
    rmqPushConsumer->setNamesrvAddr(resource->getNamesrv());
    rmqPushConsumer->setSessionCredentials(resource->getAccessKey(), resource->getSecretKey(), resource->getAccessChannel());
    rmqPushConsumer->setConsumeFromWhere(rocketmq::CONSUME_FROM_LAST_OFFSET);
    rmqPushConsumer->setConsumeThreadCount(8);
    rmqPushConsumer->subscribe(topic, tag);
    rmqPushConsumer->setMessageModel(rocketmq::MessageModel::CLUSTERING);
    rmqPushConsumer->setMaxReconsumeTimes(2);
    rmqPushConsumer->registerMessageListener(listener);
    rmqPushConsumer->start();
    
    auto producer = ProducerFactory::getRMQProducer(group);
    ASSERT_NE(producer, nullptr);
    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,std::to_string(i));
        producer->send(message);
    }

    ASSERT_EQ(SEND_NUM,producer->getEnqueueMessages()->getDataSize());
    long endTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()+240*1000L;
    while(endTime > std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()){
        if(listener->retryMsgs.size() == SEND_NUM){
            break;
        }
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }
    ASSERT_EQ(SEND_NUM,listener->retryMsgs.size());
    ASSERT_EQ(0,listener->firstMsgs.size());
    rmqPushConsumer->shutdown();
    producer->shutdown();
}

//Simulate pushconsumer consumption return null, expect that the original message was not received, and capture all messages after message retry
//// TEST(PushConsumerRetryTest, testNullConsumption){
////     int SEND_NUM = 5;
////     std::string topic = getTopic(MessageType::NORMAL, "testNullConsumption", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
////     std::string group = getGroupId("testNullConsumption");
////     std::string tag = NameUtils::getRandomTagName();
////     NullMsgListener* listener = new NullMsgListener();
////     auto rmqPushConsumer = std::make_shared<rocketmq::DefaultMQPushConsumer>(group);
////     rmqPushConsumer->setNamesrvAddr(resource->getNamesrv());
////     rmqPushConsumer->setSessionCredentials(resource->getAccessKey(), resource->getSecretKey(), resource->getAccessChannel());
////     rmqPushConsumer->setConsumeFromWhere(rocketmq::CONSUME_FROM_LAST_OFFSET);
////     rmqPushConsumer->setConsumeThreadCount(4);
////     rmqPushConsumer->subscribe(topic, tag);
////     rmqPushConsumer->setMessageModel(rocketmq::MessageModel::CLUSTERING);
////     rmqPushConsumer->setMaxReconsumeTimes(2);
////     rmqPushConsumer->registerMessageListener(listener);
////     rmqPushConsumer->start();
////     auto producer = ProducerFactory::getRMQProducer(group);
////     ASSERT_NE(producer, nullptr);
////     for(int i=0;i<SEND_NUM;i++){
////         auto message = MessageFactory::buildMessage(topic,tag,std::to_string(i));
////         producer->send(message);
////     }
////     ASSERT_EQ(SEND_NUM,producer->getEnqueueMessages()->getDataSize());
////     std::this_thread::sleep_for(std::chrono::seconds(5));
////     ASSERT_EQ(SEND_NUM,listener->retryMsgs.size());
////     ASSERT_EQ(0,listener->firstMsgs.size());
////     rmqPushConsumer->shutdown();
////     producer->shutdown();
//// }

//The normal message is sent, and after the PushConsumer retry, the retry message is expected to be consumed
TEST(PushConsumerRetryTest, testNormalTopicPushConsumerRetry){
    int SEND_NUM = 1;
    std::string topic = getTopic(MessageType::NORMAL, "testNormalTopicPushConsumerRetry", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("testNormalTopicPushConsumerRetry");
    std::string tag = NameUtils::getRandomTagName();

    NormalMsgListener* listener = new NormalMsgListener();
    auto rmqPushConsumer = std::make_shared<rocketmq::DefaultMQPushConsumer>(group);
    rmqPushConsumer->setNamesrvAddr(resource->getNamesrv());
    rmqPushConsumer->setSessionCredentials(resource->getAccessKey(), resource->getSecretKey(), resource->getAccessChannel());
    rmqPushConsumer->setConsumeFromWhere(rocketmq::CONSUME_FROM_LAST_OFFSET);
    rmqPushConsumer->setConsumeThreadCount(8);
    rmqPushConsumer->subscribe(topic, tag);
    rmqPushConsumer->setMessageModel(rocketmq::MessageModel::CLUSTERING);
    rmqPushConsumer->setMaxReconsumeTimes(2);
    rmqPushConsumer->registerMessageListener(listener);
    rmqPushConsumer->start();

    auto producer = ProducerFactory::getRMQProducer(group);
    ASSERT_NE(producer, nullptr);

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,std::to_string(i));
        producer->send(message);
    }

    ASSERT_EQ(SEND_NUM,producer->getEnqueueMessages()->getDataSize());

    long endTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()+240*1000L;
    while(endTime > std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()){
        if(listener->retryMsgs.size() == SEND_NUM && listener->firstMsgs.size() == SEND_NUM){
            break;
        }
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }

    for(auto& msg : producer->getEnqueueMessages()->getAllData()){
        ASSERT_TRUE(listener->firstMsgs.contains(msg) && listener->retryMsgs.contains(msg));
    }
    rmqPushConsumer->shutdown();
    producer->shutdown();
}


//The send order message, after the PushConsumer retry, is expected to consume the retry message, and the message consumption order and send order
TEST(PushConsumerRetryTest, testFiFoTopicPushConsumerRetry){
    int SEND_NUM = 5;
    std::string topic = getTopic(MessageType::NORMAL, "testFiFoTopicPushConsumerRetry", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("testFiFoTopicPushConsumerRetry");
    std::string tag = NameUtils::getRandomTagName();

    FIFOMsgListener* listener = new FIFOMsgListener();
    auto rmqPushConsumer = std::make_shared<rocketmq::DefaultMQPushConsumer>(group);
    rmqPushConsumer->setNamesrvAddr(resource->getNamesrv());
    rmqPushConsumer->setSessionCredentials(resource->getAccessKey(), resource->getSecretKey(), resource->getAccessChannel());
    rmqPushConsumer->setConsumeFromWhere(rocketmq::CONSUME_FROM_LAST_OFFSET);
    rmqPushConsumer->setConsumeThreadCount(4);
    rmqPushConsumer->subscribe(topic, tag);
    rmqPushConsumer->setMessageModel(rocketmq::MessageModel::CLUSTERING);
    rmqPushConsumer->setMaxReconsumeTimes(2);
    rmqPushConsumer->registerMessageListener(listener);
    rmqPushConsumer->start();

    auto producer = ProducerFactory::getRMQProducer(group);
    ASSERT_NE(producer, nullptr);

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,std::to_string(i));
        producer->sendOrderMessage(message,0);
    }

    ASSERT_EQ(SEND_NUM,producer->getEnqueueMessages()->getDataSize());

    long endTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()+240*1000L;
    while(endTime > std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count()){
        if(listener->recvMessages.size() == SEND_NUM){
            break;
        }
        std::this_thread::sleep_for(std::chrono::seconds(5));
    }
    for (int i = 0; i < SEND_NUM; i++) {
        ASSERT_EQ(std::to_string(i), listener->recvMessages[i].getBody());
    }
    rmqPushConsumer->shutdown();
    producer->shutdown();
}